client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala

```diff
@@ -1298,24 +1315,73 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
-      "DestroySlot",
-      parallelism) { case (workerInfo, (primaryLocations, replicaLocations)) =>
-      val destroy = DestroyWorkerSlots(
-        shuffleKey,
-        primaryLocations.asScala.map(_.getUniqueId).asJava,
-        replicaLocations.asScala.map(_.getUniqueId).asJava)
-      var res = requestWorkerDestroySlots(workerInfo.endpoint, destroy)
-      if (res.status != StatusCode.SUCCESS) {
-        logDebug(s"Request $destroy return ${res.status} for $shuffleKey, " +
-          s"will retry request destroy.")
-        res = requestWorkerDestroySlots(
-          workerInfo.endpoint,
-          DestroyWorkerSlots(shuffleKey, res.failedPrimarys, res.failedReplicas))
+
+    def retryDestroy(status: DestroyFutureWithStatus, currentTime: Long): Unit = {
+      status.retryTimes += 1
+      status.startTime = currentTime
+      status.future =
+        status.endpoint.ask[DestroyWorkerSlotsResponse](status.message)
+    }
+
+    val startTime = System.currentTimeMillis()
+    val futures = new util.LinkedList[DestroyFutureWithStatus]()
+    slotsToDestroy.asScala foreach { case (workerInfo, (primaryLocations, replicaLocations)) =>
+      val primaryIds = primaryLocations.asScala.map(_.getUniqueId).asJava
+      val replicaIds = replicaLocations.asScala.map(_.getUniqueId).asJava
+      val destroy = DestroyWorkerSlots(shuffleKey, primaryIds, replicaIds, mockDestroyFailure)
+      val future = workerInfo.endpoint.ask[DestroyWorkerSlotsResponse](destroy)
+      futures.add(DestroyFutureWithStatus(future, destroy, workerInfo.endpoint, 1, startTime))
+    }
+
+    val timeout = conf.rpcAskTimeout.duration.toMillis
+    var remainingTime = timeout * rpcMaxRetires
+    val delta = 50
+    while (remainingTime > 0 && !futures.isEmpty) {
+      val currentTime = System.currentTimeMillis()
+      val iter = futures.iterator()
+      while (iter.hasNext) {
+        val futureWithStatus = iter.next()
+        val message = futureWithStatus.message
+        val retryTimes = futureWithStatus.retryTimes
+        if (futureWithStatus.future.isCompleted) {
+          futureWithStatus.future.value.get match {
+            case scala.util.Success(res) =>
+              if (res.status != StatusCode.SUCCESS && retryTimes < rpcMaxRetires) {
+                logError(
+                  s"Request $message to ${futureWithStatus.endpoint} return ${res.status} for $shuffleKey $retryTimes/$rpcMaxRetires, " +
+                    "will retry.")
+                retryDestroy(futureWithStatus, currentTime)
+              } else {
+                iter.remove()
+              }
+            case scala.util.Failure(e) =>
+              if (retryTimes < rpcMaxRetires) {
+                logError(
+                  s"Request $message to ${futureWithStatus.endpoint} failed $retryTimes/$rpcMaxRetires for $shuffleKey, reason: $e, " +
+                    "will retry.")
+                retryDestroy(futureWithStatus, currentTime)
+              } else {
```

Comment:
Add a log message when the retry count exceeds the maximum retry attempts.

Reply:
done